src/disjoint_mut.rs

```diff
@@ -328,6 +342,27 @@ impl<T: ?Sized + AsMutPtr> DisjointMut<T> {
 }
 
 impl<T: AsMutPtr<Target = u8>> DisjointMut<T> {
+    /// Tell the compiler that the translated range corresponds to the given one,
+    /// and that checks don't have to be done twice.
+    ///
+    /// # Safety
+    ///
+    /// * `index` must have been `.mul(mem::size_of::<V>())`ed
+    ///   and given to [`Self::index`] or [`Self::index_mut`].
+    ///
+    /// * `slice` must be the result of the [`Self::index`] or [`Self::index_mut`] call.
+    #[inline(always)]
+    unsafe fn assume_cast_slice_len<I, V>(&self, index: I, slice: &[V])
```

Comment:
Why? I don't understand what this is doing, or why it needs to be done. Do you have justification on why we need this?

Reply:
It can eliminate a bunch of bounds checks when we index as `BitDepth16::Pixel`.

Reply:
Can you explain why better, maybe with an example in the comment? It's not clear from looking at what it is doing why that would be.

Reply:
Yeah, let me try.

Reply:
So the extra bounds checks were a result of the range possibly overflowing when we multiply it.  So the `assume` was incorrect.  I've turned it into an `assert!` now, as it still helps to replace many bounds checks on the resultant casted slice with one assert.  The assert will panic if any overflow happened in multiplying the range.  And if an overflow is impossible (e.x. the range is casted from `u32` to `usize`), then this `assert!` will be optimized out.

This is fixed in [`9bb9b2f..73f53c5`](https://github.com/memorysafety/rav1d/compare/9bb9b2fa7147263ae12d7dcbbf2a24cf7aba4c04..73f53c5deb638ec777b5306a2efd636adca18c8f).